src/OpenLoco/src/Paint/PaintTrack.cpp

```diff
@@ -27,6 +27,389 @@ namespace OpenLoco::Paint
+            },
+            std::array<int16_t, 4>{
+                reference.tunnelHeights[rotationTable[0]],
+                reference.tunnelHeights[rotationTable[1]],
+                reference.tunnelHeights[rotationTable[2]],
+                reference.tunnelHeights[rotationTable[3]],
+            },
+            std::array<SegmentFlags, 4>{
+                reference.segments[rotationTable[0]],
+                reference.segments[rotationTable[1]],
+                reference.segments[rotationTable[2]],
+                reference.segments[rotationTable[3]],
+            },
+            reference.isMergable
+        };
+    }
+
+    constexpr TrackPaintPiece kDiagonal2TPP = rotateTrackPP(kDiagonal1TPP, kDiagonal2To1Rotation);
+    constexpr TrackPaintPiece kDiagonal3TPP = rotateTrackPP(kDiagonal0TPP, kDiagonal3To0Rotation);
+
+    constexpr std::array<TrackPaintPiece, 4> kDiagonalTPP = {
+        kDiagonal0TPP,
+        kDiagonal1TPP,
+        kDiagonal2TPP,
+        kDiagonal3TPP,
+    };
+
+    constexpr std::array<std::span<const TrackPaintPiece>, 44> kTrackPaintParts = {
+        kStraightTPP,
+        kDiagonalTPP,
+    };
+
+    static void paintTrackPPMergable(PaintSession& session, const World::TrackElement& elTrack, const TrackPaintCommon& trackSession, const uint8_t rotation, const TrackPaintPiece& tpp)
+    {
+        const auto height = elTrack.baseHeight();
+        const auto heightOffset = World::Pos3{ 0,
+                                               0,
+                                               height };
+        if (elTrack.hasBridge())
+        {
+            auto newBridgeEntry = BridgeEntry(
+                height,
+                0, // straight
+                tpp.bridgeEdges[rotation],
+                tpp.bridgeQuarters[rotation],
+                elTrack.bridge(),
+                trackSession.bridgeColoursBaseImageId);
+            session.setBridgeEntry(newBridgeEntry);
+        }
+
+        const auto baseImage = trackSession.trackBaseImageId;
+
+        session.addToPlotListTrackRoad(
+            baseImage.withIndexOffset(tpp.imageIndexOffsets[rotation][0]),
+            0,
+            heightOffset,
+            tpp.boundingBoxOffsets[rotation] + heightOffset,
+            tpp.boundingBoxSizes[rotation]);
+        session.addToPlotListTrackRoad(
+            baseImage.withIndexOffset(tpp.imageIndexOffsets[rotation][1]),
+            1,
+            heightOffset,
+            tpp.boundingBoxOffsets[rotation] + heightOffset,
+            tpp.boundingBoxSizes[rotation]);
+        session.addToPlotListTrackRoad(
+            baseImage.withIndexOffset(tpp.imageIndexOffsets[rotation][2]),
+            3,
+            heightOffset,
+            tpp.boundingBoxOffsets[rotation] + heightOffset,
+            tpp.boundingBoxSizes[rotation]);
+
+        session.insertTunnels(height + tpp.tunnelHeights[rotation], trackSession.tunnelType, tpp.tunnelEdges[rotation]);
```

Comment:
mergable track pieces never have a tunnel height but might just keep this anyway.